packages/react-testing/src/root.tsx

```diff
@@ -0,0 +1,249 @@
+import * as React from 'react';
+import {render, unmountComponentAtNode} from 'react-dom';
+import {act} from 'react-dom/test-utils';
+import {
+  Arguments,
+  Props as PropsForComponent,
+  MaybeFunctionReturnType as ReturnType,
+} from '@shopify/useful-types';
+
+import {TestWrapper} from './TestWrapper';
+import {Element, Predicate} from './element';
+import {Tag, Fiber, ReactInstance, FunctionKeys} from './types';
+import {withIgnoredReactLogs} from './errors';
+
+// eslint-disable-next-line typescript/no-var-requires
+const {findCurrentFiberUsingSlowPath} = require('react-reconciler/reflection');
+
+export const connected = new Set<Root<unknown>>();
+
+export class Root<Props> {
+  get props() {
+    return this.withRoot(root => root.props);
+  }
+
+  get isDOM() {
+    return this.withRoot(root => root.isDOM);
+  }
+
+  get instance() {
+    return this.withRoot(root => root.instance);
+  }
+
+  get children() {
+    return this.withRoot(root => root.children);
+  }
+
+  get descendants() {
+    return this.withRoot(root => root.descendants);
+  }
+
+  get domNodes() {
+    return this.withRoot(root => root.domNodes);
+  }
+
+  get domNode() {
+    return this.withRoot(root => root.domNode);
+  }
+
+  private wrapper: TestWrapper<Props> | null = null;
+  private element = document.createElement('div');
+  private root: Element<Props> | null = null;
+
+  private get mounted() {
+    return this.wrapper != null;
+  }
+
+  constructor(private tree: React.ReactElement<Props>) {
+    this.mount();
+  }
+
+  perform<T>(action: () => T, {update = true} = {}): T {
+    let result!: T;
+
+    withIgnoredReactLogs(() =>
+      act(() => {
+        result = action();
+      }),
+    );
+
+    if (update) {
+      this.update();
+    }
+
+    return result;
+  }
+
+  html() {
+    return this.withRoot(root => root.html());
+  }
+
+  text() {
+    return this.withRoot(root => root.text());
+  }
+
+  is<Type extends React.ComponentType<any> | string>(
+    type: Type,
+  ): this is Root<PropsForComponent<Type>> {
+    return this.withRoot(root => root.is(type));
+  }
+
+  prop<K extends keyof Props>(key: K) {
+    return this.withRoot(root => root.prop(key));
+  }
+
+  find<Type extends React.ComponentType<any> | string>(type: Type) {
+    return this.withRoot(root => root.find(type));
+  }
+
+  findAll<Type extends React.ComponentType<any> | string>(type: Type) {
+    return this.withRoot(root => root.findAll(type));
+  }
+
+  findWhere(predicate: Predicate) {
+    return this.withRoot(root => root.findWhere(predicate));
+  }
+
+  findAllWhere(predicate: Predicate) {
+    return this.withRoot(root => root.findAllWhere(predicate));
+  }
+
+  trigger<K extends string & FunctionKeys<Props>>(
+    prop: K,
+    ...args: Arguments<Props[K]>
+  ): ReturnType<NonNullable<Props[K]>> {
+    return this.withRoot(root => root.trigger(prop, ...args));
+  }
+
+  triggerKeypath<T = unknown>(keypath: string, ...args: unknown[]) {
```

Comment:
Why isn't this just an overload of `trigger`?

Reply:
The override causes us to lose the strong typing of the return value/ arguments/ available keys, as it just flattens down to be `(key: string, ...args: unknown[]): unknown` :(

Reply:
fair fair :(